osu.Game/Overlays/Mods/ModsEffectDisplay.cs

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Extensions.LocalisationExtensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.UserInterface;
+using osu.Framework.Localisation;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Rulesets.Mods;
+using osuTK;
+
+namespace osu.Game.Overlays.Mods
+{
+    /// <summary>
+    /// Base class for displays of mods effects.
+    /// </summary>
+    public abstract class ModsEffectDisplay : Container, IHasCurrentValue<double>
+    {
+        public const float HEIGHT = 42;
+        private const float transition_duration = 200;
+
+        private readonly Box contentBackground;
+        private readonly Box labelBackground;
+        private readonly FillFlowContainer content;
+
+        public Bindable<double> Current
+        {
+            get => current.Current;
+            set => current.Current = value;
+        }
+
+        private readonly BindableWithCurrent<double> current = new BindableWithCurrent<double>();
+
+        [Resolved]
+        private OsuColour colours { get; set; } = null!;
+
+        [Resolved]
+        private OverlayColourProvider colourProvider { get; set; } = null!;
+
+        /// <summary>
+        /// Text to display in the left area of the display.
+        /// </summary>
+        protected abstract LocalisableString Label { get; }
+
+        /// <summary>
+        /// Calculates, does current value increase difficulty or decrease it.
+        /// </summary>
+        /// <param name="value">Value to calculate for. Will arrive from <see cref="Current"/> bindable once it changes.</param>
+        /// <returns>Effect of the value.</returns>
+        /// <remarks>
+        /// Default implementation compares the value with <see cref="Current"/>.<see cref="Bindable{T}.Default"/>, bigger value counts as difficulty increase.
+        /// </remarks>
+        protected virtual ModEffect CalculateEffect(double value)
+        {
+            return CalculateForSign(value.CompareTo(Current.Default));
+        }
```

Comment:
This flow is kinda weird.
1. Why is `CalculateForSign` not inlined in here?
2. How is a consumer supposed to implement this? Override `CalculateEffect` and compare it themselves?

I feel a better implementation is to remove this method, inline it in the single usage site, rename `CalculateForSign` to `GetEffectForComparison`, and make that method virtual.

Reply:
> How is a consumer supposed to implement this? Override CalculateEffect and compare it themselves?

Initially i thought that yes. But now i understand that to flip effect consumer can override `CalculateForSign` like `return base.CalculateForSign(-comparison);`. So i agree with your point.